pkg/upgrade/upgrades/schema_changes.go

```diff
@@ -210,6 +210,67 @@ func hasColumn(storedTable, expectedTable catalog.TableDescriptor, colName strin
 	return true, nil
 }
 
+// columnExists returns true if storedTable contains a column with the given
+// colName. Unlike hasColumn, it does not check that the column descriptor in
+// storedTable and expectedTable match.
+//
+// This weaker check should be used when a migration/multiple migrations
+// alter(s) the same column multiple times in order to ensure the migration(s)
+// remain(s) idempotent. Consider the following series of (sub)migrations:
+//  1. column C is first added to a table as nullable (NULL)
+//  2. column C is backfilled with non-NULL values
+//  3. column C is altered to be not nullable (NOT NULL)
```

Comment:
Can you educate me on an upgrade example where we would do >1 ddl stmts on a table?

Reply:
I think you described something [here](https://github.com/cockroachdb/cockroach/issues/91449). In that example, how did you do "step b. backfill the newly-added column" in the upgrade? Via a bunch of hard-coded `INSERT INTO` stmts?

Reply:
Yeah, so if you look at `pkg/upgrade/upgrades/system_users_role_id_migration.go`, you can see an example of this in a migration. It actually uses three separate migrations:

1. Add a `user_id` NULL column to the `system.users` table and add an index on that column
2. Use `UPDATE` statements to update all the existing rows in the table to have non-NULL `user_id` values
3. Alter the `user_id` column to be non-NULL

I believe the reason they happen in three separate migrations is that once the column is added to the table, any new inserts into the table should also write a non-NULL value to that column.

Reply:
when you say they happen in three upgrades, do you mean each is associated with a different cluster version, like
cluster version: 22 -> upgrade 1
cluster version: 24 -> upgrade 2
cluster version: 26 -> upgrade 3

If so, then what does "idempotency" here mean?

Reply:
Yes, there are 3 separate cluster versions.

Idempotence means that if upgrade 1 (adding the column to the table) is run multiple times, it should simply do nothing in any runs after the first one. Right now, if upgrade 1 were to run again it will actually return an error and this leads to an infinite upgrade retry loop.

Reply:
I see. I think I have seen similar issue before. IIRC, we simply change the upgrade stmt to something like `ADD COLUMN IF EXISTS`, `CREATE TABLE IF NOT EXISTS`, etc. This is, get idempotency from SQL commands.
Does this trick work for your case?

Reply:
The problem is if `hasColumn` is used as the `schemaExistsFn` for the operation, it will return an error before the SQL statement even runs. 

This line will cause `hasColumn` to return an error, since the descriptor after upgrade 1 will not match the final state descriptor (which is defined in `pkg/sql/catalog/systemschema/system.go`):
https://github.com/cockroachdb/cockroach/blob/c37505829d0604d34d9618077deaa7d3339256c5/pkg/upgrade/upgrades/schema_changes.go#L207-L209

This leads to this branch executing in `migrateTable`, which happens before the SQL statement would run:
https://github.com/cockroachdb/cockroach/blob/c37505829d0604d34d9618077deaa7d3339256c5/pkg/upgrade/upgrades/schema_changes.go#L110-L114

Reply:
I see. I just realized that I was asking a different question than what this PR is trying to address. Thank you for your clarification.